Backtracking/Generate-Permutations-Of-Vector.cpp

```diff
@@ -0,0 +1,98 @@
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+void initializeData(int &number_of_elements,
+                    vector<int> &vector_to_permute){
+    int number_of_elements_to_enter;
+    int current_element;
+
+    cout << "Enter the number of elements" << endl;
+    cin >> number_of_elements;
+
+    number_of_elements_to_enter = number_of_elements;
+    vector_to_permute.reserve(number_of_elements * sizeof(int));
+
+    cout << "Enter " << number_of_elements << " elements" << endl;
+
+    while(number_of_elements_to_enter > 0){
+        cin >> current_element;
+        vector_to_permute.push_back(current_element);
+        number_of_elements_to_enter--;
+    }
+}
+
+
+void initializeSolutionVector(vector<int> &current_solution,
+                              int number_of_elements) {
+    for(int i = 0; i < number_of_elements; i++ )
+        current_solution.push_back(-1);
+}
+
+/**
+ * Check if an element was used before
+ * If it was used the current solution is not valid
+ */
+bool isValid(int to_check,
+             vector<int> &current_solution){
+    for(int i = 0; i < to_check; i++)
+        if( current_solution[i] == current_solution[to_check])
+            return false;
+    return true;
+}
+
+/**
+ * Check if we used every element from vector
+ */
+bool isSolution(int current_step,
+                int number_of_elements){
+    return (current_step == (number_of_elements - 1));
+}
```

Comment:
If a function doesn't change an arguments value, use the `const` before all such arguments.